integration/ct_integration.go

```diff
@@ -0,0 +1,636 @@
+		cfg.Prefix, precertIndex, precertEntries[0].Index, leaf.Version, ts.EntryType, TimeFromMS(ts.Timestamp))
+
+	if ts.EntryType != ct.PrecertLogEntryType {
+		return fmt.Errorf("leaf[%d].ts.EntryType=%v; want PrecertLogEntryType", precertIndex, ts.EntryType)
+	}
+	if !bytes.Equal(ts.PrecertEntry.TBSCertificate, tbs) {
+		return fmt.Errorf("leaf[%d].ts.PrecertEntry differs from originally uploaded cert", precertIndex)
+	}
+	if err := stats.check(cfg, servers); err != nil {
+		return fmt.Errorf("unexpected stats check: %v", err)
+	}
+
+	// Stage 14: get an inclusion proof for the precert.
+	// Calculate leaf hash =  SHA256(0x00 | tls-encode(MerkleTreeLeaf))
+	leaf = ct.MerkleTreeLeaf{
+		Version:  ct.V1,
+		LeafType: ct.TimestampedEntryLeafType,
+		TimestampedEntry: &ct.TimestampedEntry{
+			Timestamp: precertSCT.Timestamp,
+			EntryType: ct.PrecertLogEntryType,
+			PrecertEntry: &ct.PreCert{
+				IssuerKeyHash:  sha256.Sum256(issuer.RawSubjectPublicKeyInfo),
+				TBSCertificate: tbs,
+			},
+			Extensions: precertSCT.Extensions,
+		},
+	}
+	leafData, err := tls.Marshal(leaf)
+	if err != nil {
+		return fmt.Errorf("tls.Marshal(precertLeaf)=(nil,%v); want (_,nil)", err)
+	}
+	hash := sha256.Sum256(append([]byte{merkletree.LeafPrefix}, leafData...))
+	rsp, err := pool.Pick().GetProofByHash(ctx, hash[:], sthN1.TreeSize)
+	stats.done(ctfe.GetProofByHashName, 200)
+	if err != nil {
+		return fmt.Errorf("got GetProofByHash(precertSCT, size=%d)=nil,%v", sthN1.TreeSize, err)
+	}
+	if rsp.LeafIndex != precertIndex {
+		return fmt.Errorf("got GetProofByHash(precertSCT, size=%d).LeafIndex=%d; want %d", sthN1.TreeSize, rsp.LeafIndex, precertIndex)
+	}
+	fmt.Printf("%s: Inclusion proof leaf %d @ %d -> root %d = %x\n", cfg.Prefix, precertIndex, precertSCT.Timestamp, sthN1.TreeSize, rsp.AuditPath)
+	if err := Verifier.VerifyInclusionProof(rsp.LeafIndex, int64(sthN1.TreeSize), rsp.AuditPath, sthN1.SHA256RootHash[:], leafData); err != nil {
+		return fmt.Errorf("got VerifyInclusionProof(%d,%d,...)=%v; want nil", precertIndex, sthN1.TreeSize, err)
+	}
+	if err := stats.check(cfg, servers); err != nil {
+		return fmt.Errorf("unexpected stats check: %v", err)
+	}
+
+	// Stage 15: invalid consistency proof
+	if rsp, err := pool.Pick().GetSTHConsistency(ctx, 2, 299); err == nil {
+		return fmt.Errorf("got GetSTHConsistency(2,299)=(%+v,nil); want (nil,_)", rsp)
+	}
+	stats.done(ctfe.GetSTHConsistencyName, 400)
+	fmt.Printf("%s: GetSTHConsistency(2,299)=(nil,_)\n", cfg.Prefix)
+
+	// Stage 16: invalid inclusion proof
+	wrong := sha256.Sum256([]byte("simply wrong"))
+	if rsp, err := pool.Pick().GetProofByHash(ctx, wrong[:], sthN1.TreeSize); err == nil {
+		return fmt.Errorf("got GetProofByHash(wrong, size=%d)=(%v,nil); want (nil,_)", sthN1.TreeSize, rsp)
+	}
+	stats.done(ctfe.GetProofByHashName, 400)
+	fmt.Printf("%s: GetProofByHash(wrong,%d)=(nil,_)\n", cfg.Prefix, sthN1.TreeSize)
+
+	return nil
+}
```

Comment:
Perhaps all of these stages could be broken out into functions?  This maaaaassive func   kinda blew my mind :)